File: packages/lynx-webrtc/ios/WebRTCAudioController.swift
import AVFoundation
import Foundation
import WebRTC

/// Call-audio session management, ref-counted across live peer
/// connections.
///
/// WebRTC's `RTCAudioSession` is put in manual-audio mode once (at factory
/// init) so libwebrtc never reconfigures AVAudioSession behind our back.
/// The first live peer configures `.playAndRecord` / `.voiceChat` with
/// `.defaultToSpeaker` and enables WebRTC audio; the last peer to close
/// disables it and deactivates the session. The configuration is
/// re-asserted on route changes and interruption end so another module
/// (e.g. `@sigx/lynx-audio`) flipping the shared AVAudioSession during a
/// call recovers automatically.
final class WebRTCAudioController {

    static let shared = WebRTCAudioController()

    private let queue = DispatchQueue(label: "com.sigx.webrtc.audio")
    private var livePeers = 0
    private var route: String = "speaker"
    private var observersInstalled = false

    /// Called once when the peer connection factory is created.
    func enableManualAudioMode() {
        RTCAudioSession.sharedInstance().useManualAudio = true
    }

    func retain() {
        queue.sync {
            livePeers += 1
            guard livePeers == 1 else { return }
            installObserversLocked()
            activateLocked()
        }
    }

    func release() {
        queue.sync {
            guard livePeers > 0 else { return }
            livePeers -= 1
            guard livePeers == 0 else { return }
            route = "speaker"
            deactivateLocked()
        }
    }

    /// Route call audio to `"speaker"` or `"earpiece"`. Returns an error
    /// message or nil. Devices without an earpiece keep the speaker route.
    func setRoute(_ newRoute: String) -> String? {
        guard newRoute == "speaker" || newRoute == "earpiece" else {
            return "Unknown audio route \"\(newRoute)\""
        }
        return queue.sync {
            route = newRoute
            guard livePeers > 0 else { return nil } // applied when the next call starts
            return applyRouteLocked()
        }
    }

    // MARK: - Internals (all on `queue`)

    private func activateLocked() {
        let session = RTCAudioSession.sharedInstance()
        session.lockForConfiguration()
        do {
            try session.setCategory(
                .playAndRecord,
                mode: .voiceChat,
                options: [.defaultToSpeaker, .allowBluetooth]
            )
            try session.setActive(true)
        } catch {
            NSLog("[lynx-webrtc] audio session activate failed: \(error)")
        }
        session.unlockForConfiguration()
        session.isAudioEnabled = true
        _ = applyRouteLocked()
    }

    private func deactivateLocked() {
        let session = RTCAudioSession.sharedInstance()
        session.isAudioEnabled = false
        session.lockForConfiguration()
        do {
            try session.setActive(false)
        } catch {
            // Deactivation races with other audio users — log and move on.
            NSLog("[lynx-webrtc] audio session deactivate failed: \(error)")
        }
        session.unlockForConfiguration()
    }

    private func applyRouteLocked() -> String? {
        let session = RTCAudioSession.sharedInstance()
        session.lockForConfiguration()
        defer { session.unlockForConfiguration() }
        do {
            try session.overrideOutputAudioPort(route == "speaker" ? .speaker : .none)
            return nil
        } catch {
            return "setAudioOutput failed: \(error.localizedDescription)"
        }
    }

    private func installObserversLocked() {
        guard !observersInstalled else { return }
        observersInstalled = true
        let center = NotificationCenter.default
        center.addObserver(
            self,
            selector: #selector(handleRouteChange(_:)),
            name: AVAudioSession.routeChangeNotification,
            object: nil
        )
        center.addObserver(
            self,
            selector: #selector(handleInterruption(_:)),
            name: AVAudioSession.interruptionNotification,
            object: nil
        )
    }

    @objc private func handleRouteChange(_ notification: Notification) {
        queue.async { [weak self] in
            guard let self = self, self.livePeers > 0 else { return }
            _ = self.applyRouteLocked()
        }
    }

    @objc private func handleInterruption(_ notification: Notification) {
        guard let raw = notification.userInfo?[AVAudioSessionInterruptionTypeKey] as? UInt,
              AVAudioSession.InterruptionType(rawValue: raw) == .ended else { return }
        queue.async { [weak self] in
            guard let self = self, self.livePeers > 0 else { return }
            // Re-assert the whole call configuration — the interrupting app
            // (or another module) may have changed category/active state.
            self.activateLocked()
        }
    }
}

File: packages/lynx-webrtc/android/com/sigx/webrtc/WebRTCAudioController.kt
package com.sigx.webrtc

import android.content.Context
import android.media.AudioDeviceInfo
import android.media.AudioManager
import android.os.Build
import android.util.Log

/**
 * Call-audio session management, ref-counted across live peer
 * connections: the first peer switches the [AudioManager] into
 * `MODE_IN_COMMUNICATION` and routes output to the loudspeaker (the
 * default for a hands-free call UI); the last peer to close restores
 * whatever state the app was in before.
 */
internal object WebRTCAudioController {

    private const val TAG = "SigxWebRTCAudio"

    private var livePeers = 0
    private var savedMode = AudioManager.MODE_NORMAL
    private var savedSpeakerphone = false
    private var route: String = "speaker"

    @Synchronized
    fun retain(context: Context) {
        livePeers++
        if (livePeers > 1) return
        val am = audioManager(context) ?: return
        savedMode = am.mode
        @Suppress("DEPRECATION")
        savedSpeakerphone = am.isSpeakerphoneOn
        am.mode = AudioManager.MODE_IN_COMMUNICATION
        applyRouteLocked(am)
    }

    @Synchronized
    fun release(context: Context) {
        if (livePeers == 0) return
        livePeers--
        if (livePeers > 0) return
        val am = audioManager(context) ?: return
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            am.clearCommunicationDevice()
        }
        @Suppress("DEPRECATION")
        am.isSpeakerphoneOn = savedSpeakerphone
        am.mode = savedMode
        route = "speaker"
    }

    /** Route call audio to `"speaker"` or `"earpiece"`. Persists across retain/release cycles of the same call. */
    @Synchronized
    fun setRoute(context: Context, newRoute: String): String? {
        if (newRoute != "speaker" && newRoute != "earpiece") {
            return "Unknown audio route \"$newRoute\""
        }
        route = newRoute
        if (livePeers == 0) return null // applied when the next call starts
        val am = audioManager(context) ?: return "AudioManager unavailable"
        applyRouteLocked(am)
        return null
    }

    private fun applyRouteLocked(am: AudioManager) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val wanted = if (route == "speaker") {
                AudioDeviceInfo.TYPE_BUILTIN_SPEAKER
            } else {
                AudioDeviceInfo.TYPE_BUILTIN_EARPIECE
            }
            val device = am.availableCommunicationDevices.firstOrNull { it.type == wanted }
            if (device != null) {
                if (!am.setCommunicationDevice(device)) {
                    Log.w(TAG, "setCommunicationDevice(${device.type}) refused; falling back to speakerphone flag")
                    @Suppress("DEPRECATION")
                    am.isSpeakerphoneOn = route == "speaker"
                }
                return
            }
            // Devices without an earpiece (tablets) — fall through.
        }
        @Suppress("DEPRECATION")
        am.isSpeakerphoneOn = route == "speaker"
    }

    private fun audioManager(context: Context): AudioManager? =
        context.getSystemService(Context.AUDIO_SERVICE) as? AudioManager
}
